fabric-screen-api-v1/src/main/java/net/fabricmc/fabric/impl/client/screen/ButtonList.java

```diff
@@ -32,56 +33,63 @@
 @ApiStatus.Internal
 @Environment(EnvType.CLIENT)
 public final class ButtonList extends AbstractList<ClickableWidget> {
-	private final List<Selectable> listeners;
+	private final List<Drawable> drawables;
+	private final List<Selectable> selectables;
 	private final List<Element> children;
 
-	public ButtonList(List<Selectable> listeners, List<Element> children) {
-		this.listeners = listeners;
+	public ButtonList(List<Drawable> drawables, List<Selectable> selectables, List<Element> children) {
+		this.drawables = drawables;
+		this.selectables = selectables;
 		this.children = children;
 	}
 
 	@Override
 	public ClickableWidget get(int index) {
-		return (ClickableWidget) listeners.get(translateIndex(index, false));
+		final int drawableIndex = translateIndex(drawables, index, false);
+		return (ClickableWidget) drawables.get(drawableIndex);
 	}
 
 	@Override
 	public ClickableWidget set(int index, ClickableWidget element) {
-		index = translateIndex(index, false);
-		int prevIndex = listeners.indexOf(element);
+		final int drawableIndex = translateIndex(drawables, index, false);
+		drawables.set(drawableIndex, element);
 
-		if (prevIndex >= 0 && prevIndex != index) {
-			if (prevIndex < index) index--;
-			listeners.remove(prevIndex);
-		}
-
-		int childIndex = children.indexOf(element);
+		final int selectableIndex = translateIndex(selectables, index, false);
+		selectables.set(selectableIndex, element);
 
-		if (childIndex >= 0) {
-			children.set(childIndex, element);
-		}
-
-		return (ClickableWidget) listeners.set(index, element);
+		final int childIndex = translateIndex(children, index, false);
+		return (ClickableWidget) children.set(childIndex, element);
 	}
 
 	@Override
 	public void add(int index, ClickableWidget element) {
-		index = translateIndex(index, true);
-
-		if (listeners.remove(element)) { // ensure no duplicates
+		// ensure no duplicates
+		final int duplicateIndex = drawables.indexOf(element);
+		if (duplicateIndex >= 0) {
+			drawables.remove(element);
+			selectables.remove(element);
 			children.remove(element);
-			index--;
+			if (duplicateIndex <= translateIndex(drawables, index, true)) { 
+				index--;
+			}
 		}
 
-		listeners.add(index, element);
-		this.children.add(element);
+		final int drawableIndex = translateIndex(drawables, index, true);
+		drawables.add(drawableIndex, element);
+
+		final int selectableIndex = translateIndex(selectables, index, true);
+		selectables.add(selectableIndex, element);
+
+		final int childIndex = translateIndex(children, index, true);
+		children.add(childIndex, element);
```

Comment:
The many translateIndex calls look fishy. The class is constructing a button (=ClickableWidget) view on top of the underlying wider List, besides making sure that the additions/removals remain in sync with the other collections.

This means that there should be only one direct underlying List, the one exposed by this class. `translateIndex` should translate for that List exclusively, the other collections can be manipulated arbitrarily, their indices don't have to sync up in any way unless vanilla has further requirements there.

See the existing code which uses `listeners` as the underlying List, thus using `translateIndex` for all accesses to it. Accesses to `children` don't use translation, just presence. selectables likely just needs to be handled like children was.

Reply:
I sync all lists because `drawables` determines the render order and `children` determines the tab order. I don't know exactly what `selectables` does but it has something to do with narration so I sync it just to be safe.

Reply:
You can only do this if the lists were in sync before though, otherwise the `index` becomes unrelated to the list in question.

Reply:
What you might be able to do instead is scanning for a common predecessor, but it looks like the API isn't mighty enough to handle this properly and correctly in all cases.

Reply:
```java
   protected <T extends Element & Drawable & Selectable> T addDrawableChild(T drawableElement) {
      this.drawables.add((Drawable)drawableElement);
      return this.addSelectableChild(drawableElement);
   }

   protected <T extends Element & Selectable> T addSelectableChild(T child) {
      this.children.add(child);
      this.selectables.add((Selectable)child);
      return child;
   }
```
The lists are somewhat in sync, if there are a lot of Selectables that aren't Drawable and they're before a Drawable and that Drawable is set those Selectables can end up after the Drawable.

Reply:
Per screen impl detail, if two elements are present in two of the lists, then they are guaranteed to be in the same order in relative to each other in both lists. Maybe a better approach is to use a list of any object as a backing list, and patch the list of drawable/selectable/elements as the shadows on that backing list.

Reply:
So what should I change here? I don't understand how buttons can have the same relative order without the searching for buttons done in `translateIndex`.

Reply:
you can create another list as the final reference for indexes for all types of elements, and the order in that list should be matched by other lists

Reply:
> create another list as the final reference for indexes for all types of elements

That already exists, it's `children`. It has all `Element`s in it. Then there is `selectables` which has all `Element`s that are also `Selectable` and finally `drawables` which has all `Element`s that are `Selectable` and `Drawable`.

> the order in that list should be matched by other lists

In vanilla the order of these three lists is the same, the only difference is that there are more elements in `children` than in `selectables` and more in `selectables` than in `drawables`.

Reply:
then this index call should be fine.